staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -151,11 +212,100 @@ func WithTypeAndObjectMeta(s *schema.Structural) *schema.Structural {
 	result.Properties = props
 
 	return result
-}
\ No newline at end of file
+}
+
+// estimateMinSizeJSON estimates the minimum size in bytes of the given schema when serialized in JSON.
+// minLength/minProperties/minItems is not currently taken into account, so the estimate is a little
+// smaller than the actual smallest size.
+func estimateMinSizeJSON(s *schema.Structural) int64 {
+	switch s.Type {
+	// take a comma into account for all cases since we'll generally call this function for arrays/maps
+	case "boolean":
+		// true,
+		return 5
+	case "number", "integer":
+		// 0,
+		return 2
+	case "string":
+		if s.ValueValidation != nil && s.ValueValidation.Format != "" && s.ValueValidation.Format != "byte" {
+			switch s.ValueValidation.Format {
+			case "duration":
+				return minDurationSizeJSON
+			case "date":
+				// add 1 to make room for a comma; since we're in estimateMinSizeJSON we need to account for
+				// this date being an array element
+				return dateSizeJSON + 1
+			case "date-time":
+				return minDatetimeSizeJSON
+			}
+		}
+		// "",
+		return 3
+	case "array":
+		// [],
+		return 3
+	case "object":
+		objSize := int64(3) // {},
+		// sum of all non-optional properties
+		if s.ValueValidation != nil {
+			for _, propName := range s.ValueValidation.Required {
+				prop := s.Properties[propName]
+				// add 3, 2 for quotations around the property name and 1 for the colon
+				objSize += int64(len(propName)) + estimateMinSizeJSON(&prop) + 3
+			}
+		}
+		return objSize
+	}
+	// TODO(DangerOnTheRanger): better error handling (We should never get here in normal operation)
+	return noMaxLength
+}
+
+// estimateMaxSizeJSON estimates the maximum number of elements that can fit in s considering request size
+// constraints.
+func estimateMaxSizeJSON(s *schema.Structural) int64 {
+	switch s.Type {
+	case "number", "integer":
+		// we're concerned with the theoretical largest number that could be written as a string given the
+		// request limit, not the largest 64-bit integer (which is significantly smaller), so this is the
+		// same case as a string without ValueValidation
+		return (maxRequestSizeBytes - 2)
+	case "string":
+		if s.ValueValidation != nil && s.ValueValidation.Format != "" && s.ValueValidation.Format != "byte" {
+			switch s.ValueValidation.Format {
+			case "duration":
+				return maxDurationSizeJSON
+			case "date":
+				return dateSizeJSON
+			case "date-time":
+				return maxDatetimeSizeJSON
+			}
+		}
+		// subtract 2 to account for ""
+		return (maxRequestSizeBytes - 2)
+	case "array":
+		// subtract 2 to account for [ and ]
+		return (maxRequestSizeBytes - 2) / estimateMinSizeJSON(s.Items)
+	case "object":
+		if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+			// smallest possible key ("") + colon + smallest possible value, realistically the actual keys
+			// will all vary in length
+			// TODO(DangerOnTheRanger): is there a way to calculate how many bytes unique keys will need?
```

Comment:
How about assuming keys are the quotes + 2 bytes?  If you have two bytes per key, you have no more than 2^16 possible keys which can fill about 2.8MB. The exact size would be smaller, but this is a much tighter bound than 0 or 1.

Reply:
The two-byte key size sounds good to me. I'll also adjust the tests since I think this will change a couple results.